### PR TITLE
Loadout notes

### LINKED ISSUE #7193: Allow for Notes/Description on the "Edit Loadout" Page
### Proposed change

When we create/edit a loadout, the Edit Window shows the name of the loadout, which class it applies to, the stat distribution of the loadout, and which items/mods and included in the build (including where the mods were placed using the "Show Mod Placement" button). The main feature that I believe is missing is a notes field where we can leave some comments/reminders about the build (e.g., for the stasis classes, my aspects and fragments get swapped when flipping between PVE and PVP and I might want to write down which).

I think either having a notes field that mirrors what we have on weapons/armor pieces next to the build stat line would work perfect (the red rectangle in this screenshot):

Screen Shot 2021-10-27 at 2 18 11 PM

### How does this fit into your workflow?

The ability to save, equip, and swap between builds quickly is one of DIM's best features (and one I regularly recommend to my friends). Allowing a notes field would make it substantially easier to remember specifics about a build, especially if some things need to be changed (like in the class screen or on the seasonal artifact) to get it up and running.

### PR DESCRIPTION
**Pull request by bhollis**
This adds a way to add notes to loadouts. They show up on the loadouts page, and on loadout optimizer. LO can add notes to links using `&n=` in the URL.

Fixes #7193

### CODE REVIEW

File: src/app/dim-api/dim-api-helper.ts [hunk 1]
```diff
@@ -99,12 +99,13 @@ export async function authenticatedApi<T>(config: HttpClientConfig): Promise<T>
...
 
+  let responseData: { error: string; message: string } | undefined;
   try {
-    const responseData = await response.json();
-    if (responseData.error) {
-      throw new Error(`${responseData.error}: ${responseData.message}`);
-    }
+    responseData = await response.json();
   } catch {}
+  if (responseData?.error) {
```
Comment (bhollis) [inline src/app/dim-api/dim-api-helper.ts:106]: This was just me realizing that we were swallowing the error we tried to throw from DIM API.

**Comment by nev-r** [comment]
- seems to save locally a-ok. like the add button
- not a fan of the autofocus behavior triggering on loadout re-edit. it's a weird focal priority and also swallows the escape key sheet-closing shortcut
- it's possibly unintuitive that the notes don't persist from a shared build into the saved resulting loadout
and possibly unintuitive that in order to set up LO settings and share them with a note, i have to set up, save loadout, leave this page, go to inventory/loadouts, click edit, re-optimize, and **then** they will appear in the query string

**Comment by bhollis** [comment]
> autofocus behavior
You're right, I only want it to focus when shown from the button. I'll change that.

> notes don't persist from a shared build into the saved resulting loadout
Hmm I should fix that!

> in order to set up LO settings and share them with a note
Yeah that part sucks. I was thinking of having either a share dialog where you can type notes, or making the notes area in the LO editable. Thoughts?

**Comment by nev-r** [comment]
hear me out:

:grin:

**Comment by nev-r** [comment]
share dialogue would obviously be preferable but obviously more work.
at minimum, an editable notes field would be nice until then if it's simple.

**Comment by bhollis** [comment]
OK, fixed up everything according to your suggestions.

**Review by nev-r — COMMENTED** [review]
without the autofocus behavior, i would really love to see a placeholder attribute set for the textarea. we probably have several eligible i18n strings. it appears as a thin, grey, abstract bar with almost no hint that it's even a text field, at least in my chrome.

the "cancel" behavior of the share with notes prompt is not really intuitive.

the share with notes definitely works for now but i would love to see... hm. it's really hard to say. future improvements.

**Review by nev-r — APPROVED** [review]

File: src/app/loadout-builder/LoadoutBuilder.tsx [hunk 2]
```diff
@@ -301,11 +303,15 @@ function LoadoutBuilder({
...
-  const shareBuild = () => {
-    const urlParams = new URLSearchParams({
+  const shareBuild = (notes?: string) => {
+    const p: Record<string, string> = {
       class: classType.toString(),
       p: JSON.stringify(params),
-    });
+    };
+    if (notes) {
+      p.n = notes;
```
Comment (ryan-rushton) [inline src/app/loadout-builder/LoadoutBuilder.tsx:312]: Are you using single char params because you are concerned about hitting url length limits?

**Review by ryan-rushton — APPROVED** [review]

File: src/app/loadout-builder/LoadoutBuilder.tsx (hunk 2 above)
Comment (bhollis) [reply src/app/loadout-builder/LoadoutBuilder.tsx:312]: Not really, just felt like it.

**Pull request merged**